### ISSUE TITLE
Several Linter issues

### ISSUE THREAD

**Issue by dustinmoris** [fsprojects/FSharpLint#309]
Copying the issue which was [originally raised in Ionide VS Code](https://github.com/ionide/ionide-vscode-fsharp/issues/1015):

---------

I've noticed several Linter issues:

### Constants

```fsharp
let SOME_CONSTANT_FOR_ENV_VAR = "ENV_VAR"
```

Lint: Consider changing SOME_CONSTANT_FOR_ENV_VAR to remove any underscores.

### String.Split Function

```fsharp
someVar.Split([| ','; ' ' |], StringSplitOptions.RemoveEmptyEntries)
```

![image](https://user-images.githubusercontent.com/4580714/52531826-89245900-2d13-11e9-8c5a-2f46ea757cf8.png)

The linter gets fooled by the `','`. Anything which is inside double quotes or inside single quotes is a string or character definition and probably shouldn't trigger the Linter.

### Unions

Not sure what it doesn't like here, but this is pretty common syntax in F# I thought:

![image](https://user-images.githubusercontent.com/4580714/52531849-eddfb380-2d13-11e9-82a5-ca849b88ac76.png)

### Default Property Initialisers

When I set properties to default values when initialising an object the linter mistakes it for a tuple:

```fsharp
let forwardedHeadersOptions =
            ForwardedHeadersOptions(
                ForwardedHeaders = ForwardedHeaders.XForwardedFor,
                ForwardLimit = Nullable<int>(1))
```

![image](https://user-images.githubusercontent.com/4580714/52531902-73fbfa00-2d14-11e9-835f-e300b3867952.png)

### function pattern matching

EDIT: Same as #1014 

The linter seems to confuse the pattern matching syntax with a function declaration:

![image](https://user-images.githubusercontent.com/4580714/52531916-a4dc2f00-2d14-11e9-9c7c-55cc551aa235.png)

### Comments on class members

![image](https://user-images.githubusercontent.com/4580714/52532836-da881480-2d22-11e9-81f5-3396775f82a0.png)

Otherwise works great! Got a solution which is growing in size and it's working nice!

**Comment by jgardella** [comment]
Hi @dustinmoris,

Glad FSharpLint is working well for you, and thanks for the detailed bug report. I will address all your points in turn:

**Constants**
This is a configurable setting: `LiteralNames`. By default it's set to PascalCase and no underscores. You can set a custom config to allow underscores.

**String.Split Function**
Fixed in #311.

**Unions**
There's a warning here because the `UnionDefinitionIndentation` rule expects every union case to be on its own line. We can make this configurable to allow cases on the same line.

**Default Property Initialisers**
I imagine this is being parsed as a tuple. It should also be fixed by #301.

**function pattern matching**
The linter expects the `function` keyword to be on the same line as the function definition. For example:

```
let isOne = function
    | 1 -> true
    | _ -> false
```

**Comments on class members**
Fixed in #302.

**Comment by pimbrouwers** [comment]
@jgardella With respect the `function` keyword and pattern matching, what about within pipelines?

For example, the following causes the linter to through an info currently? Despite any of my efforts to align the match expressions.

```
myCoolSequence
|> Seq.tryFind (fun n -> n.Id = 1)
|> function
      | Some myCoolItem -> Some myCoolItem
      | None -> None
```

**Comment by jgardella** [comment]
Hi @pimbrouwers.

I recently updated the rule for match indentation to be more robust. On the latest master, your example passes for me if I prefix the match clauses with 4 spaces instead of 6. If you're using Ionide it may not have these changes yet.

However there is a problem here which we need to address. None of the formatting rules have any configuration for expected number of spaces for a level of indentation; the only rule you can configure this for is the indentation rule itself. Currently all the formatting rules assume 4 spaces for indentation. We should have some global config setting for indentation, which we can refer to in all rules to be consistent.

**Comment by pimbrouwers** [comment]
@jgardella Absolutely amazing. Thanks so much not only for swift reply, but for this amazing tool!
